### PR TITLE
Add better error messages in case of some parsing problems.

### PR DESCRIPTION
**Pull request by hiker**
Some minor modification so that a user gets better error message (esp. in psyclone - instead of just "Is the Fortran correct").

### CODE REVIEW

File: src/fparser/api.py [hunk 1]
```diff
@@ -67,19 +67,18 @@
...
 
 from six import string_types
-from . import Fortran2003
 # import all Statement classes:
-from .base_classes import EndStatement, classes
-from .block_statements import *
+from .base_classes import classes
+
+from .utils import AnalyzeError
+
```
Comment (arporter) [inline src/fparser/api.py]: Tidying of imports - nice.

File: src/fparser/base_classes.py [hunk 2]
```diff
@@ -853,15 +853,16 @@ def process_subitem(self, item):
...
+            'WARNING',
+            'no parse pattern found for "%s" in %r block.'
+            % (item.get_line(), self.__class__.__name__),
+            item.span[0], item.span[1])
         logger.warning(message)
         # self.show_message(message)
         self.content.append(item)
-        #sys.exit()
+        raise AnalyzeError(message)
+        # sys.exit()
```
Comment (arporter) [inline src/fparser/base_classes.py]: I think we could delete this commented-out exit() call now.

File: src/fparser/base_classes.py [hunk 3]
```diff
@@ -853,15 +853,16 @@ def process_subitem(self, item):
         return
 
     def handle_unknown_item(self, item):
```
Comment (arporter) [inline src/fparser/base_classes.py]: Please document the interface to this routine.

File: src/fparser/parsefortran.py [hunk 4]
```diff
@@ -118,22 +118,27 @@ def put_item(self, item):
         return
 
     def parse(self):
+        '''Parses the program specified in the reader object.'''
         if self.block is not None:
             return
         try:
-            block = self.block = BeginSource(self)
+            self.block = BeginSource(self)
+        except AnalyzeError:
```
Comment (arporter) [inline src/fparser/parsefortran.py]: Catch the AnalyzeError that the parser can now raise - good.

File: src/fparser/readfortran.py [hunk 5]
```diff
@@ -810,23 +810,30 @@ def format_message(self, kind, message, startlineno, endlineno,
...
+        return self.format_message('ERROR', message, startlineno,
                                    endlineno, startcolno, endcolno)
 
-    def format_warning_message(self, message, startlineno, endlineno,
-                               startcolno=0, endcolno=-1):
-        return self.format_message('WARNING',message, startlineno,
+    def format_warning_message(self, message, startlineno=None,
+                               endlineno=None, startcolno=0, endcolno=-1):
+        '''Create a string with a warning message.'''
+        if startlineno is None:
```
Comment (arporter) [inline src/fparser/readfortran.py]: Please add a comment to explain why/how you adjust start/endlineno.

File: src/fparser/api.py [hunk 6]
```diff
@@ -216,13 +215,18 @@ def parse(input, isfree=None, isstrict=None, include_dirs = None, source_only =
...
     reader = get_reader(input, isfree, isstrict, include_dirs, source_only)
-    parser = FortranParser(reader, ignore_comments = ignore_comments)
+    parser = FortranParser(reader, ignore_comments=ignore_comments)
     parser.parse()
     if analyze:
         parser.analyze()
+    if not parser.block:
+        message = reader.format_warning_message("Parser Error")
+        raise AnalyzeError(message)
+
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/api.py]: Also, these lines are not covered by the test suite - please add a test for them.

File: src/fparser/base_classes.py (hunk 2 above)
Comment (arporter) [reply src/fparser/base_classes.py]: New code is covered but `return` statement below is now not required.

File: src/fparser/parsefortran.py (hunk 4 above)
Comment (arporter) [reply src/fparser/parsefortran.py]: This new code is covered by the test suite - good.

File: src/fparser/parsefortran.py [hunk 7]
```diff
@@ -118,22 +118,27 @@ def put_item(self, item):
...
             raise
         except:
             reader = self.reader
             while reader is not None:
-                message = reader.format_message('FATAL ERROR',
-                                                'while processing line',
-                                                reader.linecount, reader.linecount)
+                message = \
+                    reader.format_message('FATAL ERROR',
+                                          'while processing line',
```
Comment (arporter) [inline src/fparser/parsefortran.py]: This code is not covered but as it has only been tidied for pep8 I won't request it.

File: src/fparser/readfortran.py (hunk 5 above)
Comment (arporter) [reply src/fparser/readfortran.py]: In fact, these lines aren't covered by the test suite either. Please can you add/modify tests such that they are.

**Comment by hiker** [comment]
Hi,
thanks for the feedback. My pull request might need a bit of an explanation: no, I did not just remove the line that had no test case. It turns out that when this line would be triggered, the original parser message was lost. So instead I am now raising an exception, so that the error message can properly be handled.
One additional question: what is the best way to check the coverage? I can see see the link in this pull request after I push to my repo, but I am sure there's a command-line-way of checking that before I push.
Thanks!
   Joerg

**Comment by rupertford** [comment]
For command line checking I use

` py.test --cov-report term-missing --cov fparser.xxx`
 where xxx is the python module name.

It requires an extension to py.test which I believe is installed thus ...

`sudo pip install pytest-cov`

**Comment by arporter** [comment]
My preferred output is html as it makes cross-referencing with the changes (displayed in a github PR) much easier:

    py.test --cov-report html ...

File: src/fparser/readfortran.py [hunk 8]
```diff
@@ -797,14 +797,15 @@ def format_message(self, kind, message, startlineno, endlineno,
...
+                                        len(self.source_lines))+1):
+            if i == 0 and not self.source_lines:
                 break
             linenostr = '%5d:' % (i)
-            if i==endlineno:
+            if i == endlineno:
                 sourceline = self.source_lines[i-1]
                 l0 = linenostr+sourceline[:startcolno]
-                if endcolno==-1:
+                if endcolno == -1:
```
Comment (arporter) [inline src/fparser/readfortran.py]: Tidying for pep8 - thanks!

File: src/fparser/readfortran.py (hunk 5 above)
Comment (arporter) [reply src/fparser/readfortran.py]: Thanks for adding the docstring. However, the new lines of code are not covered. Also, what happens if `startlineno=None` and `self.linecount < 2`?

File: src/fparser/tests/test_parser.py [hunk 9]
```diff
@@ -869,3 +870,14 @@ def test_invalid_type_bound_array_access():  # pylint: disable=invalid-name
     with pytest.raises(ValueError) as excinfo:
         _ = parse(Call, 'call an_array(idx)%)')
     assert "with Call pattern failed" in str(excinfo)
+
+
+def test_analyze_errors():
```
Comment (arporter) [inline src/fparser/tests/test_parser.py]: New test - great!

**Comment by hiker** [comment]
Sorry, wasn't entirely ready - since I needed to check coverage (which I didn't know how to do) - plus the link on the web interface didn't work for me this time (I get: "Source Not Available   The file "/home/travis/virtualenv/python2.7.13/lib/python2.7/site-packages/fparser/readfortran.py" isn't available on github. Either it's been removed, or the repo root directory needs to be updated.).

But thanks for the feedback, but will be a few days before I can get to that I guess.

File: src/fparser/readfortran.py (hunk 5 above)
Comment (hiker) [reply src/fparser/readfortran.py]: This confuses me a bit ... while I agree that in my original pull request I had not handled that case, but I've fixed that, see: https://github.com/stfc/fparser/pull/55/files:
 +      if startlineno is None:
+            startlineno = self.linecount-2
+            if startlineno < 0:
+                startlineno = 1

The change above is marked as 'outdated' - I assume there was a mis-click somewhere ;)

**Comment by arporter** [comment]
Oops - sorry for jumping the gun in that case. I shall wait till you tell me you're ready next time! That message about missing source is because the test suite now runs as two sub-jobs in Travis - one for Python 2.7 and one for Python 3.something. Things are currently set-up so that the source browser will find the source for the Python3.something tests but not the 2.7. I haven't investigated to see whether it's possible to get it to work for both.

**Comment by hiker** [comment]
Don't worry about the link, since you gave me the info how to do it myself I don't need them anymore ;)

**Comment by hiker** [comment]
OK, next try. I was quite confused by the 'new code not covered by test case' (format_warning or so function), since it was covered when I tested it. But turhs out that when I cleaned up some code based on your initial comments, I actually removed the code that would have triggered the format_warning call in the first place - so it was safe to just remove the code.
And while I usually avoid adding more new stuff in a pull-request fix, there were some really low hanging fruits and I added two or three easy tests.
The travis errors are apparently only python3 related, so I assume I can ignore that?

Hope that's all now :)

Thanks!
   Joerg

**Comment by arporter** [comment]
Thanks @hiker, I'm afraid ignoring failing tests (even for Python 3) is not an option :-) Please could you fix and then I'll review...

**Comment by hiker** [comment]
Oops - misunderstanding. I thought that python3 is not yet supported (and as part of an upcoming port someone had set up python3 test ... never realised that they were running all the time). OK, I tried to fix it (though not sure how to trigger the python3 tests locally yet).

**Pull request merged**